### ISSUE TITLE
Extended JDTLS support

### ISSUE THREAD

**Issue by eruizc-dev** [neovim/nvim-lspconfig#1067]
![image](https://user-images.githubusercontent.com/58179604/125474999-8456d09c-8ff2-4dca-8e12-55d2a6e734ad.png)

### Clarification

This is something I'd like to do and contribute. I'm creating an issue to see if there is interest from the community and it stays within this project's scope before I spend time on something that won't be merged.

### Issue

JDTLS is a non-conventional language server, I've been using it for 3 months and it works fine but sometimes errors pop up, code generations always use incorrect indentation and I know there's a lot of missing features.

I know there's an amazing plugin ([nvim-jdtls](https://github.com/mfussenegger/nvim-jdtls)) that takes care of this, but it more than just a language server client including testing, debugging, bytecode inspection and more...

There's also an issue on how it's launched, the command is huge (you gotta spin up jvm with lots of arguments) and there's no convention. Also, the default cmd has no Lombok support which is a **must** for writing modern java.

### Proposal

 - Add some extended client capabilities
 - Autodetect if user installed jdtls via [AUR](https://aur.archlinux.org/packages/java-language-server/) or [JDTLS-LAUNCHER](https://github.com/eruizc-dev/jdtls-launcher)
 - Allow Lombok setup with default cmd
 - Add better documentation

**Comment by mjlbach** [comment]
I'm in favor of anything that does not substantially add to the complexity of the jdtls configuraiton in nvim-lspconfig, however, I am not in favor of anything that is too opinionated (does adding lombok support result in strange behavior for those not using lombok?) or anything that is tailored to a specific installation method (I would consider the AUR to be an unreliable source generally).

Re: extending client capabilities which capabilities were you thinking of? If we advertise something core does not support it may result in undefined behavior from the language server.

I really want to push people to use/improve nvim-jdtls, is there an issue with not using the additional features it affords?

**Comment by mfussenegger** [comment]
> but it more than just a language server client including testing, debugging, bytecode inspection and more...

All of that functionality is based on the language server. The debug-adapter is an extension to the language server, and doesn't work without it. The bytecode inspection is using the language server to retrieve the classpath of your project. Besides - having that functionality in it has no overhead at runtime. Most of it also doesn't have an overhead at import time, because it lives in separate modules.

> There's also an issue on how it's launched, the command is huge (you gotta spin up jvm with lots of arguments) and there's no convention. Also, the default cmd has no Lombok support which is a must for writing modern java.

One of the reasons I ended up letting users create their own config is so that they can easily extend the java/jvm options - which for example lets them add lombok support or configure the heap size. 

Besides that, @mjlbach and me were actually talking about how we could streamline the setup to be closer to what lspconfig is doing.

> Add better documentation

Would love to hear what's missing from the nvim-jdtls documentation

**Comment by eruizc-dev** [comment]
> extending client capabilities which capabilities were you thinking of?

All the necessary ones to prevent `No delegateCommandHandler` error and all code actions that can be resolved with no more 
than 10 lines of code to reduce complexity

> does adding lombok support result in strange behavior for those not using lombok?

No, Lombok is a library that provides annotations to generate implicit methods such as getters, setters, constructors and others. You need to add a javaagent to jdtls execution command for it to detect the existence of these implicit methods.

> Both the launcher in AUR and the JDTLS-LAUNCHER ignore that you probably want to set a different workspace folder per project.

Agree, but the location of this directory does not affect the language server at all. I think there has to be a standard and provide an option to change it.

> Would love to hear what's missing from the nvim-jdtls documentation

I'm glad that you saw this, I wanted to reach you at some point too.

I've been talking via discord with other neovim users when I had to set up JDTLs. nvim-jdtls had all the features I wanted but it wasn't easy when you know nothing. Everyone pointed me to [this video](https://youtu.be/E-MvQC04Cbo) but it still took me like 3 hours to set up. I've seen many people resort to COC for this reason, you can't beat a `:CocInstall`.

When I had everything set up my vimrc became a mess. Having a separate own code_actions function and the start_or_attach working different to lspconfig's setup() function, I ended up with lots of conditionals and autocmds.

Because I need Java for work and I didn't want to spend all working day setting up everything, I created jdtls and decided to go lspconfig only, and talking with others they found it useful too reducing jdtls setup time to about 5 minutes.

My ultimate goal here is to reduce the time it takes to set up jdtls and try make use of all its features. What I see here is:
 - JDTLS is lacking an official installer and launcher, 
 - Some of the features that nvim-jdtls provides could be upstreamed into lspconfig so people that don't need anything extra doesn't have to install and configure an aditional plugin
 - nvim-jdtls does an excellent job but the API and documentation could be improved

I'm willing to help in all 3 cases, I just want to know what's the best way of doing so

**Comment by mjlbach** [comment]
> Some of the features that nvim-jdtls provides could be upstreamed into lspconfig so people that don't need anything extra doesn't have to install and configure an additional plugin

I think the issue here is, I don't use Java, and I can't really support people asking specific questions about Java. We're getting to a fairly massive number of language servers, and when I want to make tree-side changes that are not reg-exable it takes an unreasonable amount of time. I'd like to move more and more to create a common framework for language specific extensions like the nvim-jdtls model. Lspconfig should be a decent fallback that implements *only* things on the LSP specification. I've been somewhat lax about this historically, but as language specific plugins emerge I'm in favor of removing anything we've added in lspconfig that is off-spec.

> JDTLS is lacking an official installer and launcher

I'm hoping package managers add it to their repos, or that people prefer the prebuilt binaries which should be easy to put somewhere on the path. I think `JDTLS-LAUNCHER` would need, at minimum, to remove all usage of elevated privileges (don't install things into non-user directories) for me to consider adding it in the installation instructions. If people are really having issues getting it installed, the `LspInstall` plugin should likely be updated.

**Comment by mfussenegger** [comment]
> All the necessary ones to prevent No delegateCommandHandler error and all code actions that can be resolved with no more than 10 lines of code to reduce complexity

By default the only command that eclipse.jdt.ls sends that isn't part of the core spec is the `java.apply.workspaceEdit`, that should already be handled here: https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/jdtls.lua#L107


Any of the other extensions require an opt-in via a setting in the `init_options`. Supporting them you'd likely end up requiring a dedicated code-action function as well. There is a reason I had to do that in nvim-jdtls, see https://github.com/neovim/neovim/issues/12326 / https://github.com/neovim/neovim/pull/14115 - There is work to make that unnecessary in the future.



> When I had everything set up my vimrc became a mess. Having a separate own code_actions function and the start_or_attach working different to lspconfig's setup() function, I ended up with lots of conditionals and autocmds.

Can you elaborate why you needed conditionals and lots of autocmds?  Using both lspconfig setup and the autocmd for jdtls should be possible without any extra conditionals.
The `code_action` function from jdtls is even a drop-in replacement for `vim.lsp.buf.code_action`. (And otherwise you could work with 2 different `on_attach` hooks.)


> nvim-jdtls does an excellent job but the API and documentation could be improved


It would help getting more insights into where you got stuck, or with which part you struggled. "could be improved" is too generic and not actionable feedback.

**Comment by eruizc-dev** [comment]
> I'm hoping package managers add it to their repos, or that people prefer the prebuilt binaries which should be easy to put somewhere on the path.

Thank you for the feedback, I plan to push this idea to Eclipse for an official launcher or improve mine i.

> I don't use Java, and I can't really support people asking specific questions about Java.

My plans did not extend outside the existing `jdtls.lua` file, simply updating the multiple handlers and providing much better settings. Anyways, I see why you wouldn't want to have things outside spec.

> By default the only command that eclipse.jdt.ls sends that isn't part of the core spec is the java.apply.workspaceEdit

Yes but still the current way of handling it causes the error to appear when I organize imports or remove variables. I think it should either be removed or completed.

> Any of the other extensions require an opt-in via a setting in the init_options.

Yes, this is pretty much what I'd like to support by default. As far as I've seen it should be possible to implement at least a few and the issue you linked was something I also considered and it's cool to see that there's a discussion about that. I mainly opened this issue to know what the scope for this project is and know where I should push for changes.

> "could be improved" is too generic and not actionable feedback

I didn't want to go into details because I don't think it's the right place, do you mind if I open an issue on your repo with full details?

**Comment by mjlbach** [comment]
>   My plans did not extend outside the existing jdtls.lua file, simply updating the multiple handlers and providing much better settings. Anyways, I see why you wouldn't want to have things outside spec.

Sure! If there's a way to improve the handlers that's not implemented in lspconfig go for it :) I expect though that it will require duplicating a lot of functionality in nvim-jdlts (esp around code actions), but I'll defer to @mfussenegger on this one

**Comment by mfussenegger** [comment]
>>Any of the other extensions require an opt-in via a setting in the init_options.

> Yes, this is pretty much what I'd like to support by default


This reads as "I'd like lspconfig to do what nvim-jdtls already does, because ..."
Duplicating these efforts is not a good use of our time. The idea was and is for `lspconfig` to contain a set of default configurations for various language servers. Any "not part of core specification" extensions are out of scope. If it wasn't for that - I could've integrated the nvim-jdtls functionality in lspconfig to begin with.

I'd rather spend the time on addressing the *because* reasons.

Given that many servers provide functionalities that are beyond the core LSP spec, we expect more language specific plugins to pop-up - and there are already several of them.

If you've concrete suggestions on how to make the extensions feel more aligned, so that any `nvim-lsp-LANG-X` doesn't feel foreign but provides a similar setup experience, let us know.


> I didn't want to go into details because I don't think it's the right place, do you mind if I open an issue on your repo with full details?

Please go for it.

**Comment by eruizc-dev** [comment]
Sorry for the late response but I had to think a little before answering!

> This reads as "I'd like lspconfig to do what nvim-jdtls already does, because ..."

My original plan was to extract some of nvim-jdtls functionality and include it in lspconfig; I wanted to split the problem of having to set up two plugins at the same time into setting one up first and then the other. After talking with you and thinking about it I came to the same conclusion as yours that it will simply take too much time for the benefits.

To close this issue I'd like to make a contribution including a documentation update. Whenever someone wants to set up a language server the first thing they will look at is CONFIG.md file and it's far from providing enough information to be able to set up at least the basics.

I will also put effort into nvim-jdtls which seems to be the best solution.

Thank you both @mjlbach and @mfussenegger and sorry if I took much time of yours with this discussion, I wanted to reduce jdtls setup time and didn't know where to start!
